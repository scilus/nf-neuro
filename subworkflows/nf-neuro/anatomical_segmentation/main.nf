// ** Importing modules from nf-neuro ** //
include { SEGMENTATION_FASTSEG       } from '../../../modules/nf-neuro/segmentation/fastseg/main'
include { SEGMENTATION_FREESURFERSEG } from '../../../modules/nf-neuro/segmentation/freesurferseg/main'
include { SEGMENTATION_SYNTHSEG      } from '../../../modules/nf-neuro/segmentation/synthseg/main'

params.run_synthseg = false

workflow ANATOMICAL_SEGMENTATION {

    // ** Two input channels for the segmentation processes since they are using   ** //
    // ** different image files. Supplying an empty channel for the one that isn't ** //
    // ** relevant will make the workflow run the appropriate module.              ** //
    take:
        ch_image            // channel: [ val(meta), [ image ] ]
        ch_freesurferseg    // channel: [ val(meta), [ aparc_aseg, wmparc ] ], optional
        ch_lesion           // channel: [ val(meta), [ lesion ] ], optional
        ch_fs_license       // channel: [ val[meta], [ fs_license ] ], optional

    main:

        ch_versions = Channel.empty()

        if ( params.run_synthseg ) {
            // ** Freesurfer synthseg segmentation ** //
            SEGMENTATION_SYNTHSEG (
                ch_image
                    .join(ch_lesion, remainder: true)
                    .map{ it[0..1] + [it[2] ?: []] }
                    .combine(ch_fs_license)
            )
            ch_versions = ch_versions.mix(SEGMENTATION_SYNTHSEG.out.versions.first())

            // ** Setting outputs ** //
            wm_mask = SEGMENTATION_SYNTHSEG.out.wm_mask
            gm_mask = SEGMENTATION_SYNTHSEG.out.gm_mask
            csf_mask = SEGMENTATION_SYNTHSEG.out.csf_mask
            wm_map = SEGMENTATION_SYNTHSEG.out.wm_map
            gm_map = SEGMENTATION_SYNTHSEG.out.gm_map
            csf_map = SEGMENTATION_SYNTHSEG.out.csf_map
            seg = SEGMENTATION_SYNTHSEG.out.seg
            aparc_aseg = SEGMENTATION_SYNTHSEG.out.aparc_aseg
            resample = SEGMENTATION_SYNTHSEG.out.resample
            volume = SEGMENTATION_SYNTHSEG.out.volume
            qc_score = SEGMENTATION_SYNTHSEG.out.qc_score
        }

        else {
            // ** FSL fast segmentation ** //
            SEGMENTATION_FASTSEG (
                ch_image
                    .join(ch_lesion, remainder: true)
                    .map{ it[0..1] + [it[2] ?: []] }
            )
            ch_versions = ch_versions.mix(SEGMENTATION_FASTSEG.out.versions.first())

            // ** Setting outputs ** //
            wm_mask = SEGMENTATION_FASTSEG.out.wm_mask
            gm_mask = SEGMENTATION_FASTSEG.out.gm_mask
            csf_mask = SEGMENTATION_FASTSEG.out.csf_mask
            wm_map = SEGMENTATION_FASTSEG.out.wm_map
            gm_map = SEGMENTATION_FASTSEG.out.gm_map
            csf_map = SEGMENTATION_FASTSEG.out.csf_map
            seg = Channel.empty()
            aparc_aseg = Channel.empty()
            resample = Channel.empty()
            volume = Channel.empty()
            qc_score = Channel.empty()


            // ** Freesurfer segmentation ** //
            SEGMENTATION_FREESURFERSEG (
                ch_freesurferseg
                    .join(ch_lesion, remainder: true)
                    .map{ it[0..2] + [it[3] ?: []] }
            )
            ch_versions = ch_versions.mix(SEGMENTATION_FREESURFERSEG.out.versions.first())

            // ** Setting outputs ** //
            wm_mask = wm_mask.mix( SEGMENTATION_FREESURFERSEG.out.wm_mask )
            gm_mask = gm_mask.mix( SEGMENTATION_FREESURFERSEG.out.gm_mask )
            csf_mask = csf_mask.mix( SEGMENTATION_FREESURFERSEG.out.csf_mask )
        }

    emit:
        wm_mask    = wm_mask                     // channel: [ val(meta), [ wm_mask ] ]
        gm_mask    = gm_mask                     // channel: [ val(meta), [ gm_mask ] ]
        csf_mask   = csf_mask                    // channel: [ val(meta), [ csf_mask ] ]
        wm_map     = wm_map                      // channel: [ val(meta), [ wm_map ] ]
        gm_map     = gm_map                      // channel: [ val(meta), [ gm_map ] ]
        csf_map    = csf_map                     // channel: [ val(meta), [ csf_map ] ]
        seg        = seg                         // channel: [ val(meta), [ seg ] ]
        aparc_aseg = aparc_aseg                  // channel: [ val(meta), [ aparc_aseg ] ]
        resample   = resample                    // channel: [ val(meta), [ resample ] ]
        volume     = volume                      // channel: [ val(meta), [ volume ] ]
        qc_score   = qc_score                    // channel: [ val(meta), [ qc_score ] ]

        versions  = ch_versions                  // channel: [ versions.yml ]
}
