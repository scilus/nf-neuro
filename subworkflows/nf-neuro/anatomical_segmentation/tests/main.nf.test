nextflow_workflow {

    name "Test Subworkflow ANATOMICAL_SEGMENTATION"
    script "../main.nf"
    workflow "ANATOMICAL_SEGMENTATION"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/anatomical_segmentation"

    tag "segmentation"
    tag "segmentation/fastseg"
    tag "segmentation/freesurferseg"
    tag "segmentation/synthseg"

    tag "load_test_data"

        setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "T1w.zip", "freesurfer_nifti.zip", "freesurfer.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("anatomical_segmentation - fslfast") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        freesurfer_nifti: it.simpleName == "freesurfer_nifti"
                        freesurfer: it.simpleName == "freesurfer"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz", checkIfExists: true)
                    ]
                }
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("anatomical_segmentation - freesurferseg") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        freesurfer_nifti: it.simpleName == "freesurfer_nifti"
                        freesurfer: it.simpleName == "freesurfer"
                    }
                input[0] = Channel.empty()
                input[1] = ch_split_test_data.freesurfer_nifti.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/aparc+aseg.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/wmparc.nii.gz", checkIfExists: true)
                    ]
                }
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("anatomical_segmentation - both") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        freesurfer_nifti: it.simpleName == "freesurfer_nifti"
                        freesurfer: it.simpleName == "freesurfer"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz", checkIfExists: true)
                    ]
                }
                input[1] = ch_split_test_data.freesurfer_nifti.map{
                    test_data_directory -> [
                        [ id:'test' ], // meta map
                        file("\${test_data_directory}/aparc+aseg.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/wmparc.nii.gz", checkIfExists: true)
                    ]
                }
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("anatomical_segmentation - synthseg") {
        config "./nextflow_synthseg.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        freesurfer_nifti: it.simpleName == "freesurfer_nifti"
                        freesurfer: it.simpleName == "freesurfer"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz", checkIfExists: true)
                    ]
                }
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                input[3] = ch_split_test_data.freesurfer.map{
                    test_data_directory -> [
                        file("\${test_data_directory}/license.txt", checkIfExists: true)
                    ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    file(workflow.out.wm_mask.get(0).get(1)).name,
                    file(workflow.out.gm_mask.get(0).get(1)).name,
                    file(workflow.out.csf_mask.get(0).get(1)).name,
                    file(workflow.out.wm_map.get(0).get(1)).name,
                    file(workflow.out.gm_map.get(0).get(1)).name,
                    file(workflow.out.csf_map.get(0).get(1)).name,
                    file(workflow.out.seg.get(0).get(1)).name,
                    workflow.out.versions
                ).match()}
            )
        }
    }
}
