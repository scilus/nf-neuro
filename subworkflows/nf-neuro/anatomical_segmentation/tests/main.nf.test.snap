{
    "anatomical_segmentation - freesurferseg": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_wm.nii.gz:md5,0ed4a06defc512b1b898d6e0c3ef678a"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_gm.nii.gz:md5,a363d6207419b08346a7c05c6525430e"
                    ]
                ],
                "10": [
                    
                ],
                "11": [
                    "versions.yml:md5,18dfcb72144006d4f3e2afc759d2ee6b"
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_csf.nii.gz:md5,3dbcb3eefb507bedab91fb80ed636f36"
                    ]
                ],
                "3": [
                    
                ],
                "4": [
                    
                ],
                "5": [
                    
                ],
                "6": [
                    
                ],
                "7": [
                    
                ],
                "8": [
                    
                ],
                "9": [
                    
                ],
                "aparc_aseg": [
                    
                ],
                "csf_map": [
                    
                ],
                "csf_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_csf.nii.gz:md5,3dbcb3eefb507bedab91fb80ed636f36"
                    ]
                ],
                "gm_map": [
                    
                ],
                "gm_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_gm.nii.gz:md5,a363d6207419b08346a7c05c6525430e"
                    ]
                ],
                "qc_score": [
                    
                ],
                "resample": [
                    
                ],
                "seg": [
                    
                ],
                "versions": [
                    "versions.yml:md5,18dfcb72144006d4f3e2afc759d2ee6b"
                ],
                "volume": [
                    
                ],
                "wm_map": [
                    
                ],
                "wm_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_wm.nii.gz:md5,0ed4a06defc512b1b898d6e0c3ef678a"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-12T03:20:09.242971787"
    },
    "anatomical_segmentation - synthseg": {
        "content": [
            "test__mask_wm.nii.gz",
            "test__mask_gm.nii.gz",
            "test__mask_csf.nii.gz",
            "test__map_wm.nii.gz",
            "test__map_gm.nii.gz",
            "test__map_csf.nii.gz",
            "test__seg.nii.gz",
            [
                "versions.yml:md5,95e31db072c7bb2505128bd08beaecf4"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-30T16:59:25.733839079"
    },
    "anatomical_segmentation - fslfast": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_wm.nii.gz:md5,98d75f5c0ddd2b7f2bfff8fbf1877e34"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_gm.nii.gz:md5,a840f08c5b90cb351d3272d8cab6f00c"
                    ]
                ],
                "10": [
                    
                ],
                "11": [
                    "versions.yml:md5,f098240b7d7f6df9a2a43215c5828669"
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_csf.nii.gz:md5,f22037987a38e5132ccbb2887ee71842"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_wm.nii.gz:md5,c7d666c66baea2132d70be38017c6901"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_gm.nii.gz:md5,6dbc993e43f6efcdd6a3a097df2b4574"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_csf.nii.gz:md5,dcc224f10f31ee306491ebbdcfbe5bfe"
                    ]
                ],
                "6": [
                    
                ],
                "7": [
                    
                ],
                "8": [
                    
                ],
                "9": [
                    
                ],
                "aparc_aseg": [
                    
                ],
                "csf_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_csf.nii.gz:md5,dcc224f10f31ee306491ebbdcfbe5bfe"
                    ]
                ],
                "csf_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_csf.nii.gz:md5,f22037987a38e5132ccbb2887ee71842"
                    ]
                ],
                "gm_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_gm.nii.gz:md5,6dbc993e43f6efcdd6a3a097df2b4574"
                    ]
                ],
                "gm_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_gm.nii.gz:md5,a840f08c5b90cb351d3272d8cab6f00c"
                    ]
                ],
                "qc_score": [
                    
                ],
                "resample": [
                    
                ],
                "seg": [
                    
                ],
                "versions": [
                    "versions.yml:md5,f098240b7d7f6df9a2a43215c5828669"
                ],
                "volume": [
                    
                ],
                "wm_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_wm.nii.gz:md5,c7d666c66baea2132d70be38017c6901"
                    ]
                ],
                "wm_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_wm.nii.gz:md5,98d75f5c0ddd2b7f2bfff8fbf1877e34"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-12T03:17:48.731576565"
    },
    "anatomical_segmentation - both": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_wm.nii.gz:md5,0ed4a06defc512b1b898d6e0c3ef678a"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_wm.nii.gz:md5,98d75f5c0ddd2b7f2bfff8fbf1877e34"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_gm.nii.gz:md5,a363d6207419b08346a7c05c6525430e"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_gm.nii.gz:md5,a840f08c5b90cb351d3272d8cab6f00c"
                    ]
                ],
                "10": [
                    
                ],
                "11": [
                    "versions.yml:md5,18dfcb72144006d4f3e2afc759d2ee6b",
                    "versions.yml:md5,f098240b7d7f6df9a2a43215c5828669"
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_csf.nii.gz:md5,3dbcb3eefb507bedab91fb80ed636f36"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_csf.nii.gz:md5,f22037987a38e5132ccbb2887ee71842"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_wm.nii.gz:md5,c7d666c66baea2132d70be38017c6901"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_gm.nii.gz:md5,6dbc993e43f6efcdd6a3a097df2b4574"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_csf.nii.gz:md5,dcc224f10f31ee306491ebbdcfbe5bfe"
                    ]
                ],
                "6": [
                    
                ],
                "7": [
                    
                ],
                "8": [
                    
                ],
                "9": [
                    
                ],
                "aparc_aseg": [
                    
                ],
                "csf_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_csf.nii.gz:md5,dcc224f10f31ee306491ebbdcfbe5bfe"
                    ]
                ],
                "csf_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_csf.nii.gz:md5,3dbcb3eefb507bedab91fb80ed636f36"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_csf.nii.gz:md5,f22037987a38e5132ccbb2887ee71842"
                    ]
                ],
                "gm_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_gm.nii.gz:md5,6dbc993e43f6efcdd6a3a097df2b4574"
                    ]
                ],
                "gm_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_gm.nii.gz:md5,a363d6207419b08346a7c05c6525430e"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_gm.nii.gz:md5,a840f08c5b90cb351d3272d8cab6f00c"
                    ]
                ],
                "qc_score": [
                    
                ],
                "resample": [
                    
                ],
                "seg": [
                    
                ],
                "versions": [
                    "versions.yml:md5,18dfcb72144006d4f3e2afc759d2ee6b",
                    "versions.yml:md5,f098240b7d7f6df9a2a43215c5828669"
                ],
                "volume": [
                    
                ],
                "wm_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_wm.nii.gz:md5,c7d666c66baea2132d70be38017c6901"
                    ]
                ],
                "wm_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_wm.nii.gz:md5,0ed4a06defc512b1b898d6e0c3ef678a"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_wm.nii.gz:md5,98d75f5c0ddd2b7f2bfff8fbf1877e34"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-12T03:22:31.039354334"
    }
}